walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -36,50 +39,171 @@ public int compare(Event a, Event b) {
-    List<TimeRange> possibleMeetingTimes = new ArrayList<TimeRange>();
+    return times;
+  }
+
+  /**
+   * Returns a change log of the number of available optional attendees over time. First part of a
+   * sweep-line algorithm.
+   *
+   * @param optionalAttendeesFreeTimes mapping of all attendees to their free times
+   */
+  private ArrayList<Map.Entry<Integer, Integer>> getChanges(
+      HashMap<String, ArrayList<TimeRange>> optionalAttendeesFreeTimes) {
+    TreeMap<Integer, Integer> changes = new TreeMap<Integer, Integer>();
+    for (Map.Entry e : optionalAttendeesFreeTimes.entrySet()) {
+      for (TimeRange time : (ArrayList<TimeRange>) e.getValue()) {
+        changes.put(time.start(), changes.getOrDefault(time.start(), 0) + 1);
+        changes.put(time.end(), changes.getOrDefault(time.end(), 0) - 1);
+      }
+    }
+    return new ArrayList<Map.Entry<Integer, Integer>>(changes.entrySet());
+  }
+
+  /**
+   * Updates optimalMeetingTimes based on current meeting time's attendance.
+   *
+   * @param time current meeting time range
+   * @param bestAttendance best attendance of optional attendees seen so far
+   * @param currAttendance attendance optional attendees of current meeting
+   * @param optionalAttendeesFreeTimes all attendees and their free times
+   * @param minTime minimum meeting time
+   */
+  private int updateOptimalTimes(
+      TimeRange time,
+      int bestAttendance,
+      int currAttendance,
+      ArrayList<TimeRange> optimalMeetingTimes,
+      long minTime) {
+    if (time.duration() >= minTime) {
+
+      // Clear out all former optimal meeting times. They aren't the most optimal anymore.
+      if (currAttendance > bestAttendance) {
+        bestAttendance = currAttendance;
+        optimalMeetingTimes.clear();
+      }
+      if (currAttendance == bestAttendance) {
+        optimalMeetingTimes.add(time);
+      }
+    }
+    return bestAttendance;
+  }
+
+  /**
+   * Returns all meeting times that satisfy all mandatory attendees of request. Sorted in ascending
+   * order.
+   *
+   * @param events all events to be considered
+   * @param minTime minimum length of time for meeting
+   * @param mandatoryAttendees everyone who needs to attend this meeting
+   */
+  private ArrayList<TimeRange> getMandatoryAttendeesMeetingTimes(
+      Collection<Event> events, long minTime, HashSet<String> mandatoryAttendees) {
+    ArrayList<Event> relevantEvents =
+        getRelevantEvents(mandatoryAttendees, new ArrayList<Event>(events));
+    ArrayList<TimeRange> possibleMeetingTimes = new ArrayList<TimeRange>();
 
     // Need to check this so we don't access out of bounds when we add first gap.
-    if (events.isEmpty()) {
+    if (relevantEvents.isEmpty()) {
       addIfLongEnough(
-          TimeRange.fromStartEnd(0, END_OF_DAY, true), possibleMeetingTimes, request.getDuration());
+          TimeRange.fromStartEnd(0, TimeRange.END_OF_DAY, true), possibleMeetingTimes, minTime);
```

Comment:
What happens if `addIfLongEnough` doesn't actually add? I feel like something should happen (in all of these cases), but you're not handling it at all.

I expected this function to return a `bool` representing whether or not the value was added, because I expect the failure to add the event to have consequences.

Reply:
l don't think anything needs to happen. The purpose of that function was just to avoid code duplication (3 times). lt just compares a time range with how long the meeting needs to be and if it's too short, doesn't bother adding it.